metrics/internal/cache/pv.go

```diff
@@ -0,0 +1,151 @@
+package cache
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"sync"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/watch"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog"
+)
+
+// Cache metadata for all PV with annotation - pv.kubernetes.io/provisioned-by: *.rbd.csi.ceph.com
+
+var _ cache.Store = &PersistentVolumeStore{}
+
+// PersistentVolumeStore implements the k8s.io/client-go/tools/cache.Store
+// interface. It stores persistent volume CSI attributes
+type PersistentVolumeStore struct {
+	Mutex sync.RWMutex
+	// Store is a map of PV UID to PersistentVolumeAttributes
+	Store map[types.UID]PersistentVolumeAttributes
+}
+
+type PersistentVolumeAttributes struct {
+	PersistentVolumeName           string
+	PersistentVolumeClaimName      string
+	PersistentVolumeClaimNamespace string
+	ImageName                      string
+	Pool                           string
+}
+
+func NewPersistentVolumeStore() *PersistentVolumeStore {
+	return &PersistentVolumeStore{
+		Store: map[types.UID]PersistentVolumeAttributes{},
+	}
+}
+
+// Add inserts to the PersistentVolumeStore.
+func (p *PersistentVolumeStore) Add(obj interface{}) error {
+	pv, ok := obj.(*corev1.PersistentVolume)
+	if !ok {
+		return fmt.Errorf("unexpected object of type %T", obj)
+	}
+
+	provisioner := pv.Annotations["pv.kubernetes.io/provisioned-by"]
+	if !strings.Contains(provisioner, ".rbd.csi.ceph.com") {
```

Comment:
Good to have a constant for the CSI driver name in this case?

Reply:
It's not used anywhere else. So literal should be fine.
Also, this is throwaway code until Ceph implementation is available. So I'm not worried much about other aspects.